### PR TITLE
Runtime-agnostic HTTP handler

### PR DESCRIPTION
**Pull request by baoshan**
Follow up to https://github.com/octokit/auth-oauth-user-client.js/pull/1#issuecomment-878291468

This commit distills the [original](https://github.com/octokit/oauth-app.js/blob/52ba35a928291a6ac44d25de93ca5f0cb0fe9847/src/middleware/node/middleware.ts) Node.js/Express middleware into a [runtime agnostic HTTP handler](https://github.com/baoshan/oauth-app.js/blob/92869e374ad90876f861043a002f5c14b01d2653/src/middleware/handler.ts) which operates on a generalized request/response interface rather than `IncomingMessage`.

This handler hopefully can better support different runtimes (Cloudflare/Deno/AWS Lambda) with the same behavior, so the client authentication strategy can also be backend-or-platform-agnostic.

The [new](https://github.com/baoshan/oauth-app.js/blob/92869e374ad90876f861043a002f5c14b01d2653/src/middleware/node/index.ts) Node.js/Express middleware relies on the generalized HTTP handler. Test file is untouched to respect the current behavior.

`fromentries` dependency is removed since Node.js v10 is out of maintanence.

### CODE REVIEW

**Comment by wolfy1339** [comment]
> fromentries dependency is removed since Node.js v10 is out of maintanence.
We still support NodeJS v10 for now, until our next major version bump

File: src/middleware/types.ts
Comment (gr2m) [inline src/middleware/types.ts:11]: can we rename `text` to `body`? That would fit the convention we have in https://github.com/octokit/endpoint.js which also has the concept of "generic request options":
image

See also [`RequestOptions` type in `@octokit/types](https://github.com/octokit/types.ts/blob/579aa75c8cf8c90997cc52f0f3145da8bad232d6/src/RequestOptions.ts)

File: src/middleware/types.ts
Comment (gr2m) [inline src/middleware/types.ts:5]: i'm not 100% sure about this API surface. I guess you are trying to stay close to the web platform [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) object? We can leave it as is because I don't have a better idea right now, maybe it will grow on me 🤷🏼

File: src/middleware/types.ts
Comment (gr2m) [reply src/middleware/types.ts:11]: I'm conflicted about text vs body, property vs getter vs function ... It's just internal so it's easy enough to change later I guess

**Review by gr2m — COMMENTED** [review]
Okay this is quite a lot 😁 I like your approach, but I think you've done two steps at once. I think it would have been a better approach to first create a Cloudflare Worker middleware and get it merged. And once we have that we could have looked into abstractions. But with only two use cases I fear that an abstraction might be premature and cause work work down the path, I'd probably looked into yet another middleware that is as different as possible to Node's and Cloudflare.

But that's okay now, I'm just thinking out loud, let's be pragmatic about it. For now, can you please address my comments?

Could you also please add `README.md` files with some explanations in `src/middleware` (what is this folder / how to build your own middleware) and `src/midleware/node` (how to use the node/express middleware)

File: src/middleware/types.ts
Comment (baoshan) [reply src/middleware/types.ts:11]: When I see `bodyString` somewhere (even locally declared) then my OCD starts enslaving me to use  `bodyJSON` wherever the body is `JSON` parsed (for request) or to be serialized (for response).

I’d like to stick to a common standard (even a flawed one as long as I’m not the one to blame).

- Node.js is not a good reference since it exposes an evented API (`data/end/etc.`).
- Express defines an untyped `.body` property on the request, the concrete type relies on the body parser (or `undefined` if there’s no one). I guess that’s the (partial) reason it is named `body`.
- [Response in Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response) names its first parameter `body` which accepts a string among `FormData/ReadableStream/etc.` but does not accept an object. I think this explicitness is good. The Request object is explicit in exposing a `text()` async function.

This is why I explicitly named the property `text` in both `OctokitRequest` and `OctokitResponse`. I do respect the maintainers’ final decision: if `text` is inappropriate, just let me know.

Comment (gr2m) [reply src/middleware/types.ts:11]: let's leave it as is for now, thanks 👍🏼

**Comment by gr2m** [comment]
I resolved the conflict and workedaround a typescript error. I had to force-push the changes so watch out if you continue working on it. This also triggered the full CI on the PR

File: src/middleware/README.md
Comment (gr2m) [inline src/middleware/README.md:89]: thanks for the README, this is great!

File: src/middleware/node/parse-request.ts
Comment (gr2m) [inline src/middleware/node/parse-request.ts:2]: good catch, thank you for fixing it!

**Review by gr2m — APPROVED** [review]
Thanks a lot, I really like how readable the node middleware code is now!

To share the bigger picture of middlewares in Octokit land: we do have the concept several `@octokit/*` modules, some of which depend on each other.

- [`@octokit/webhooks`](https://github.com/octokit/webhooks.js/tree/b09d16484882c7af0d3ba6d9335b33f74658f59d#createnodemiddleware)
- [`@octokit/oauth-app`](https://github.com/octokit/oauth-app.js/tree/7a7225b200e9f43f9fe6ef1565740ae1e9d0afd2#middlewares)
- [`@octokit/app`](https://github.com/octokit/app.js/tree/c7e2b10f771c9a4b0b1c60effad98a9bab25da16#middlewares) - depends on `@octokit/webhooks`s and `@octokit/oauth-app`'s middleware
- [`octokit`](https://github.com/octokit/octokit.js/#readme) - uses `@octokit/app`'s middleware

I think once we are happy with the implementation in `@octokit/oauth-app`, it would make sense to look to move the abstract middleware to handler requests/responses into its own module, so that the other packages can utilize it. No need to worry about it now, just wanted to mention it so you better understand the context of where I'm coming from

**Review by Lekir1 — COMMENTED** [review]
@cferuz83

**Pull request merged**
